File: Python/Stack/ValidParenthesis.py
class Stack:
    def __init__(self):
        self.stack = []  # List to store the elements of the stack

    def push(self, data):
        self.stack.append(data)  # Add an element to the top of the stack

    def pop(self):
        if not self.is_empty():
            return self.stack.pop()  # Remove and return the top element of the stack
        else:
            print("Stack is empty.")
            return None

    def is_empty(self):
        return len(self.stack) == 0  # Check if the stack is empty

def is_valid(s):
    stack = Stack()  # Create a stack object

    for ch in s:
        if ch == '(' or ch == '{' or ch == '[':
            stack.push(ch)  # If the character is an opening parenthesis, push it onto the stack
        else:
            if ch == ')':
                if stack.is_empty() or stack.pop() != '(':  # If the character is a closing parenthesis ')', check if the stack is empty or the top of the stack does not match the corresponding opening parenthesis '('
                    return False  # Unbalanced parentheses
            elif ch == '}':
                if stack.is_empty() or stack.pop() != '{':  # Similar check for closing parenthesis '}'
                    return False  # Unbalanced parentheses
            elif ch == ']':
                if stack.is_empty() or stack.pop() != '[':  # Similar check for closing parenthesis ']'
                    return False  # Unbalanced parentheses

    return stack.is_empty()  # The parentheses are balanced if the stack is empty after processing all characters

s = "{}()[]"
if is_valid(s):
    print("Valid parentheses")
else:
    print("Invalid parentheses")

File: C/Stack/ValidParenthesis.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

// Node structure for the stack
typedef struct {
    char data;
    struct Node* next;
} Node;

// Stack structure
typedef struct {
    Node* top;
} Stack;

// Function to create a new stack
Stack* createStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    stack->top = NULL;
    return stack;
}

// Function to push an element onto the stack
void push(Stack* stack, char data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = stack->top;
    stack->top = newNode;
}

// Function to pop an element from the stack
char pop(Stack* stack) {
    // Check if the stack is empty
    if (stack->top == NULL) {
        printf("Stack is empty.\n");
        return '\0';
    }

    char data = stack->top->data;
    Node* temp = stack->top;
    stack->top = stack->top->next;
    free(temp);
    return data;
}

// Function to check if the stack is empty
bool isEmpty(Stack* stack) {
    return stack->top == NULL;
}

// Function to validate parentheses in a string
bool isValid(char* s) {
    Stack* stack = createStack();

    // Iterate through each character in the string
    for (int i = 0; s[i] != '\0'; i++) {
        char ch = s[i];
        if (ch == '(' || ch == '{' || ch == '[') {
            // If the character is an opening parenthesis, push it onto the stack
            push(stack, ch);
        } else {
            if (ch == ')') {
                // If the character is a closing parenthesis, check if the stack is empty or the top of the stack does not match the corresponding opening parenthesis
                if (isEmpty(stack) || pop(stack) != '(') {
                    return false;
                }
            }
            if (ch == '}') {
                if (isEmpty(stack) || pop(stack) != '{') {
                    return false;
                }
            }
            if (ch == ']') {
                if (isEmpty(stack) || pop(stack) != '[') {
                    return false;
                }
            }
        }
    }

    // Check if the stack is empty after processing all characters
    return isEmpty(stack);
}

int main() {
    char s[] = "{}()[]";
    if (isValid(s)) {
        printf("Valid parentheses\n");
    } else {
        printf("Invalid parentheses\n");
    }
    return 0;
}
